packages/models-tree/src/models-tree/ModelsTreeDefinition.ts

```diff
@@ -740,40 +774,72 @@ async function createInstanceKeyPathsFromTargetItems({
   }
   const hierarchyConfig = props.hierarchyConfig ?? defaultHierarchyConfiguration;
   const idsCache = props.idsCache ?? new ModelsTreeIdsCache(imodelAccess, hierarchyConfig);
-  const ids = {
-    models: new Array<Id64String>(),
-    categories: new Array<Id64String>(),
-    subjects: new Array<Id64String>(),
-    elements: new Array<Id64String | ElementsGroupInfo>(),
-  };
-  await Promise.all(
-    targetItems.map(async (key) => {
-      if ("parent" in key) {
-        ids.elements.push(key);
-      } else if (await imodelAccess.classDerivesFrom(key.className, "BisCore.Subject")) {
-        ids.subjects.push(key.id);
-      } else if (await imodelAccess.classDerivesFrom(key.className, "BisCore.Model")) {
-        ids.models.push(key.id);
-      } else if (await imodelAccess.classDerivesFrom(key.className, "BisCore.SpatialCategory")) {
-        ids.categories.push(key.id);
-      } else {
-        ids.elements.push(key.id);
-      }
-    }),
-  );
-  const elementBlocks: Array<Array<Id64String | ElementsGroupInfo>> = [];
-  const elementsLength = ids.elements.length;
-  const blockSize = Math.ceil(elementsLength / Math.ceil(elementsLength / 5000));
-  for (let i = 0; i < ids.elements.length; i += blockSize) {
-    elementBlocks.push(ids.elements.slice(i, i + blockSize));
-  }
-
-  return collect(
-    merge(
-      from(ids.subjects).pipe(mergeMap((id) => createSubjectInstanceKeysPath(id, idsCache))),
-      from(ids.models).pipe(mergeMap((id) => createModelInstanceKeyPaths(id, idsCache))),
-      from(ids.categories).pipe(mergeMap((id) => createCategoryInstanceKeyPaths(id, idsCache))),
-      from(elementBlocks).pipe(mergeMap((block) => createGeometricElementInstanceKeyPaths(imodelAccess, idsCache, hierarchyConfig, block))),
+  return lastValueFrom(
+    from(targetItems).pipe(
+      releaseMainThreadOnItemsCount(2000),
+      mergeMap(async (key): Promise<{ key: string; type: number } | { key: ElementsGroupInfo; type: 0 }> => {
+        if ("parent" in key) {
+          return { key, type: 0 };
+        }
+
+        const subjectCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.Subject");
+        const isSubject = subjectCheck instanceof Promise ? await subjectCheck : subjectCheck;
+        if (isSubject) {
+          return { key: key.id, type: 1 };
+        }
```

Comment:
Does the `instanceof Promise` check have any value over simply doing `const isSubject = await imodelAccess.classDerivesFrom(key.className, "BisCore.Subject");`?

Reply:
Well, it's tricky to say. If it has then it's very minor. I ran the test 50 times (one with the `instanceof` check and one without) and compared results, implementation with `instanceof` check was about 100ms faster on average. Then i did it again and it seemed to be just about 10 ms faster.
Since its not obvious that this helps I've removed the `instanceof` check.